test/package-url.spec.js

```diff
@@ -169,6 +169,42 @@ describe('PackageURL', function () {
                 testInvalid(paramName)
             })
         })
+
+        it('should normalize params', () => {
+            assert.strictEqual(
+                new PackageURL('type', '%21', 'name').toString(),
+                'pkg:type/!/name'
+            )
```

Comment:
This behavior seems wrong. It's very surprising that the strings being passed to the constructor are percent decoded before being incorporated into the PURL.

```
> var a = new PackageURL("generic", "%2525", "name"); a.toString();
'pkg:generic/%2525/name'
> var b = new PackageURL(a.type, a.namespace, a.name); b.toString();
'pkg:generic/%25/name'
> var c = new PackageURL(b.type, b.namespace, b.name); c.toString();
Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at Object.normalizeNamespace [as normalize] (/purl/repo/src/normalize.js:16:25)
    at new PackageURL (/purl/repo/src/package-url.js:50:39)
```
(from 2.0)

Reply:
@matt-phylum  I'm still noodling on normalization behavior of the constructor. We may be able to tweak it to how URLSearchParams does it (it must have some encoding detection). The PR is in a draft state to indicate not ready just yet. The constructor is a place of validation and some normalization. It may be that we normalize and validate but don't store the normalized value. Still digging into it.